File: src/my_py_pkg/my_py_pkg/my_first_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

# classe MyNode que herda a classe Node do ros
class MyNode(Node):
    
    def __init__(self): # construtor da classe
        super().__init__("py_test")  # invoca o contrutor do Node e nomeia o nó de py_test
        self.counter = 0 # cria um contador
        self.get_logger().info("Hello ROS2 em python")
        self.create_timer(0.5, self.timer_callback) #chama a função timer_callback a cada 0.5 segundo

    def timer_callback(self):
        self.counter += 1
        self.get_logger().info("Hello " + str(self.counter))


def main(args=None):
    rclpy.init(args=args) # precisa estar em qualquer nó pois inicaliza a comunicação do ROS2 
    node = MyNode() # cria o objeto MyNode  
    rclpy.spin(node) # faz com que o programa continue rodando e so pare com crtl c
    rclpy.shutdown() # finaliza a comunicação

if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/my_first_node.cpp
#include "rclcpp/rclcpp.hpp"

class MyNode: public rclcpp::Node // cria a classe MyNode que herda a classe Node no rclcpp
{
public:
    MyNode(): Node("cpp_test") // invoca o construtor da classe Node e o nomeia o nó de cpp_test
    {
        RCLCPP_INFO(this->get_logger(), "Hello cpp Node!");
        timer = this->create_wall_timer(std::chrono::seconds(1), 
            std::bind(&MyNode::timerCallback, this)); // cria timer de 1 segundo que chama a função timerCallback
    }

private:

    void timerCallback()
    {
        counter++;
        RCLCPP_INFO(this->get_logger(), "Hello %d", counter);
    }

    rclcpp::TimerBase::SharedPtr timer; // declara um timer do ros2 para shared pointer chamado timer
    int counter = 0;
};

int main(int argc, char **argv){
    
    rclcpp::init(argc, argv); // inicializa a comunicação do ros
    auto node = std::make_shared<MyNode>(); // cria um shared pointer pra o nó MyNode
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}   
